### PR TITLE
ContentLengthVerifier Update

### PR DESCRIPTION
**Pull request by carolinacass**
Motivation:
Currently, swift-nio-http2 ignores content length when handling a response to a HEAD request and handling a 304 response.

Modifications:
Changing ContentLengthVerifier and any functions called

### CODE REVIEW

File: Tests/NIOHTTP2Tests/ConnectionStateMachineTests.swift [hunk 1]
```diff
@@ -3001,6 +3001,156 @@ class ConnectionStateMachineTests: XCTestCase {
...
+        // The server responds
+        assertSucceeds(self.server.sendHeaders(streamID: streamTwo, headers: responseHeaders, isEndStreamSet: false))
+        assertSucceeds(self.client.receiveHeaders(streamID: streamTwo, headers: responseHeaders, isEndStreamSet: false))
+
+        // Send in 1 byte over one frame
+        assertStreamError(type: HTTP2ErrorCode.protocolError, self.server.sendData(streamID: streamTwo, contentLength: 1, flowControlledBytes: 1, isEndStreamSet: true))
+        assertStreamError(type: HTTP2ErrorCode.protocolError, self.client.receiveData(streamID: streamTwo, contentLength: 1, flowControlledBytes: 1, isEndStreamSet: true))
+    }
+
+    func testSimpleServerPush_forPushHeadRequest() {
```
Comment (Lukasa) [inline Tests/NIOHTTP2Tests/ConnectionStateMachineTests.swift:3130]:
```suggestion
    func testPushHeadRequest() {
```

File: Tests/NIOHTTP2Tests/ConnectionStateMachineTests.swift [hunk 2]
```diff
@@ -3001,6 +3001,156 @@ class ConnectionStateMachineTests: XCTestCase {
...
+
+        // Server can push right away
+        assertSucceeds(self.server.sendPushPromise(originalStreamID: streamOne, childStreamID: streamTwo, headers: requestHeaders))
+        assertSucceeds(self.client.receivePushPromise(originalStreamID: streamOne, childStreamID: streamTwo, headers: requestHeaders))
+
+        // The server responds
+        assertSucceeds(self.server.sendHeaders(streamID: streamTwo, headers: responseHeaders, isEndStreamSet: false))
+        assertSucceeds(self.client.receiveHeaders(streamID: streamTwo, headers: responseHeaders, isEndStreamSet: false))
+
+        // Send in 0 bytea over one frame
```
Comment (Lukasa) [inline Tests/NIOHTTP2Tests/ConnectionStateMachineTests.swift:3151]:
```suggestion
        // Send in 0 bytes over one frame
```

File: Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift [hunk 3]
```diff
@@ -19,83 +19,98 @@ import NIOHPACK
...
         // Int.max + 1
         headers = HPACKHeaders([("Host", "apple.com"), ("content-length", "9223372036854775809"), ("User-Agent", "myCoolClient/1.0")])
-        XCTAssertThrowsError(try ContentLengthVerifier(headers)) { error in
+        XCTAssertThrowsError(try ContentLengthVerifier(headers, requestMethod: nil)) { error in
             XCTAssertTrue(error is NIOHTTP2Errors.ContentLengthHeaderMalformedValue)
         }
     }
+
+    func testContentLengthVerifier_whenRequestMethod() throws {
+        let headers = HPACKHeaders([("Host", "apple.com"), ("content-length", "1834"), ("User-Agent", "myCoolClient/1.0"), (":status", "304")])
```
Comment (Lukasa) [inline Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift:103]: Strictly `:status` has to come first here.

File: Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift [hunk 4]
```diff
@@ -19,83 +19,98 @@ import NIOHPACK
...
+        XCTAssertNoThrow(try ContentLengthVerifier(headers, requestMethod: "GET"))
+        let verifier = try assertNoThrowWithValue(try ContentLengthVerifier(headers, requestMethod: "GET"))
+        XCTAssertEqual(0, verifier.expectedContentLength)
+    }
+
+
+    func testContentLengthVerifier_whenRequestMethodofHead() throws {
```
Comment (Lukasa) [inline Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift:110]:
```suggestion
    func testContentLengthVerifier_whenRequestMethodIsHead() throws {
```

File: Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift [hunk 5]
```diff
@@ -19,83 +19,98 @@ import NIOHPACK
...
 
         // Int.max + 1
         headers = HPACKHeaders([("Host", "apple.com"), ("content-length", "9223372036854775809"), ("User-Agent", "myCoolClient/1.0")])
-        XCTAssertThrowsError(try ContentLengthVerifier(headers)) { error in
+        XCTAssertThrowsError(try ContentLengthVerifier(headers, requestMethod: nil)) { error in
             XCTAssertTrue(error is NIOHTTP2Errors.ContentLengthHeaderMalformedValue)
         }
     }
+
+    func testContentLengthVerifier_whenRequestMethod() throws {
```
Comment (Lukasa) [inline Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift:102]: This should probably say something about 304, not the request method.

**Review by Lukasa — COMMENTED** [review]
Thanks for this! I've left some notes in the diff.

File: Sources/NIOHTTP2/ContentLengthVerifier.swift [hunk 6]
```diff
@@ -82,3 +92,14 @@ extension ContentLengthVerifier: CustomStringConvertible {
...
+
+extension HPACKHeaders {
+    func getPseudoHeader(name: String) -> String? {
+        for (fieldName, fieldValue, _) in self {
+            if name == fieldName {
+                return fieldValue
+            }
+        }
+        return nil
+    }
```
Comment (glbrntt) [inline Sources/NIOHTTP2/ContentLengthVerifier.swift:104]: I don't think we need this function: we have `headers.first(name:)` which does the same.

File: Sources/NIOHTTP2/StreamStateMachine.swift [hunk 7]
```diff
@@ -284,13 +284,16 @@ extension HTTP2StreamStateMachine {
             switch self.state {
             case .idle(.client, localWindow: let localWindow, remoteWindow: let remoteWindow):
                 let targetState: State
-                let localContentLength = validateContentLength ? try ContentLengthVerifier(headers) : .disabled
+//                since we are a client and this is a request, server sending a response
+                let localContentLength = validateContentLength ? try ContentLengthVerifier(headers,requestMethod: nil) : .disabled
```
Comment (glbrntt) [inline Sources/NIOHTTP2/StreamStateMachine.swift:288]:
```suggestion
                let localContentLength = validateContentLength ? try ContentLengthVerifier(headers, requestMethod: nil) : .disabled
```

File: Sources/NIOHTTP2/ConnectionStateMachine/ConnectionStreamsState.swift [hunk 8]
```diff
@@ -77,10 +77,11 @@ struct ConnectionStreamState {
     /// - parameters:
     ///     - streamID: The ID of the pushed stream.
     ///     - remoteInitialWindowSize: The initial window size of the remote peer.
+    ///     - requestVerb: HTTP Request Method used for different purposes with a web server
```
Comment (Lukasa) [inline Sources/NIOHTTP2/ConnectionStateMachine/ConnectionStreamsState.swift:80]: I think we can make this clearer: we can say "the HTTP method used on the request".

File: Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift [hunk 9]
```diff
@@ -19,83 +19,98 @@ import NIOHPACK
...
 
         // Int.max + 1
         headers = HPACKHeaders([("Host", "apple.com"), ("content-length", "9223372036854775809"), ("User-Agent", "myCoolClient/1.0")])
-        XCTAssertThrowsError(try ContentLengthVerifier(headers)) { error in
+        XCTAssertThrowsError(try ContentLengthVerifier(headers, requestMethod: nil)) { error in
             XCTAssertTrue(error is NIOHTTP2Errors.ContentLengthHeaderMalformedValue)
         }
     }
+
+    func testContentLengthVerifier_whenRequestMethodIs304() throws {
```
Comment (Lukasa) [inline Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift:102]:
```suggestion
    func testContentLengthVerifier_whenResponseStatusIs304() throws {
```

File: Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift [hunk 10]
```diff
@@ -19,83 +19,98 @@ import NIOHPACK
...
+        let headers = HPACKHeaders([(":status", "304"), ("Host", "apple.com"), ("content-length", "1834"), ("User-Agent", "myCoolClient/1.0")])
+        XCTAssertNoThrow(try ContentLengthVerifier(headers, requestMethod: "GET"))
```
Comment (Lukasa) [inline Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift:104]: We don't need this line, the line below covers this.

File: Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift [hunk 11]
```diff
@@ -19,83 +19,98 @@ import NIOHPACK
...
+        let verifier = try assertNoThrowWithValue(try ContentLengthVerifier(headers, requestMethod: "GET"))
+        XCTAssertEqual(0, verifier.expectedContentLength)
+    }
+
+
+    func testContentLengthVerifier_whenRequestMethodIsHead() throws {
+        let headers = HPACKHeaders([("Host", "apple.com"), ("content-length", "1834"), ("User-Agent", "myCoolClient/1.0")])
+        XCTAssertNoThrow(try ContentLengthVerifier(headers, requestMethod: "HEAD"))
```
Comment (Lukasa) [inline Tests/NIOHTTP2Tests/ContentLengthVerifierTests.swift:112]: We don't need this line, the line below covers it.

**Review by Lukasa — APPROVED** [review]

**Comment by Lukasa** [comment]
Can you run `scripts/generate_linux_tests.rb` please?

**Pull request merged**